Database_MySql.cs for the NetworkZones addon in the latest Version

Assambra opened the pull request.

Assambra committed:
Fixed some typing Error in Database_MySql.cs

If we work without the Database_Compat.cs this will give us an Error

Assambra committed:
Merge branch 'master' of https://github.com/Assambra/ummorpg_mysql

Assambra committed:
Database_MySql.cs for NetworkZones_2018-03-16 Addon

Assambra committed:
missing information in the readme.txt

paulpach commented:
Thank you for this,  but this duplicates the entire class.

A better way is to add a partial class that adds the stuff needed for network zones.   I don't want to maintain 2 copies of the same thing.

paulpach reviewed (changes requested):
If I merged this pull request I would have 2 copies of the same class.  This is a maintenance burden.

Could you refactor it so that you simply have a partial class that adds the stuff needed for networkzones?

Assambra committed:
The NetworkZones_MySql.cs is now a partial Database class, simple add into the uMMORPG Addons Folder

Assambra commented:
Add NetworkZones_MySql as a partial database class and removed the Database_MySql.cs from the Addons

paulpach commented on Addons/NetworkZones/NetworkZones_MySql.cs:
```diff
@@ -0,0 +1,107 @@
+        //   current time otherwise
+        // -> it uses the ISO 8601 standard format
+        string onlineString = DateTime.UtcNow.ToString("s");
+        ExecuteNonQueryMySql("DELETE FROM zones_online");
```
These 2 lines are not atomic.  Thus if a different zone queries the table between these line it will seem the zone is offline.

Instead,  use a single [insert on duplicate](https://dev.mysql.com/doc/refman/5.7/en/insert-on-duplicate.html)

paulpach commented on Addons/NetworkZones/NetworkZones_MySql.cs:
```diff
@@ -0,0 +1,107 @@
+
+    public static void SaveCharacterScenePath(string characterName, string scenePath)
+    {
+        ExecuteNonQueryMySql("REPLACE INTO character_scene VALUES (@character, @scenepath)",
```
Replace has some issues,  I prefer to use [insert on duplicate](https://dev.mysql.com/doc/refman/5.7/en/insert-on-duplicate.html),   it is faster and has less surprises when using foreign keys

paulpach commented on Addons/NetworkZones/NetworkZones_MySql.cs:
```diff
@@ -0,0 +1,107 @@
+        object obj = ExecuteScalarMySql("SELECT online FROM characters WHERE name=@name", new SqlParameter("@name", characterName));
+        if (obj != null)
+        {
+            string online = (string)obj;
```
online is not a string in my addon,  I changed it to a TIMESTAMP,  this will throw an invalid cast exception.

paulpach commented on Addons/NetworkZones/NetworkZones_MySql.cs:
```diff
@@ -0,0 +1,107 @@
+
+        ExecuteNonQueryMySql(@"
+        CREATE TABLE IF NOT EXISTS zones_online (
+            online VARCHAR(255) NOT NULL
```
Change this to TIMESTAMP

Assambra replied:
this is the zone online check to stay consistend with the addon i let it as VARCHAR(). But you point me in the right direction here i have to change it
```#c
public static bool IsCharacterOnlineAnywhere(string characterName)
    {
        object obj = ExecuteScalarMySql("SELECT online FROM characters WHERE name=@name", new SqlParameter("@name", characterName));
        if (obj != null)
        {
            string online = (string)obj;
            if (online != "")
            {
                DateTime time = DateTime.Parse(online);
                double elapsedSeconds = (DateTime.UtcNow - time).TotalSeconds;
                float saveInterval = ((NetworkManagerMMO)NetworkManager.singleton).saveInterval;
                //UnityEngine.Debug.Log("datetime=" + time + " elapsed=" + elapsedSeconds + " saveinterval=" + saveInterval);
                return elapsedSeconds < saveInterval * 2;
            }
        }
        return false;
    }
```

paulpach replied:
I suggested to vis to change this,  he will "look into it".   But even if he does not change it, I don't care about doing it exactly the same way vis did it.  I would rather do it right,  plus it is not like the fields need to maintain compatibility,  the method signatures are the ones that matter.

paulpach reviewed (changes requested):
Please fix the use of online and make SaveMainZoneOnlineTime atomic

paulpach commented on Addons/NetworkZones/NetworkZones_MySql.cs:
```diff
@@ -0,0 +1,107 @@
+    {
+        ExecuteNonQueryMySql(@"
+        CREATE TABLE IF NOT EXISTS character_scene (
+            `character` VARCHAR(16) NOT NULL, 
```
add a foreign key to characters with this field with on delete cascade.   That way if the character is deleted,  this table is automatically cleaned up as well.

Assambra replied:
FOREIGN KEY(`character`) ON DELETE CASCADE ON UPDATE CASCADE,
Maybe name can changed also in future, GM or something else

paulpach replied:
```
FOREIGN KEY(`character`)
                REFERENCES characters(name)
                ON DELETE CASCADE ON UPDATE CASCADE
```

See examples in my class

paulpach reviewed (changes requested):
Add foreign key with cascade delete

paulpach commented on Addons/NetworkZones/NetworkZones_MySql.cs:
```diff
@@ -0,0 +1,107 @@
+        object obj = ExecuteScalarMySql("SELECT online FROM zones_online");
+        if (obj != null)
+        {
+            string online = (string)obj;
```
change this one to use DateTime  if you change the field to a TIMESTAMP

Assambra committed:
Now using TIMESTAMP instead of a VARCHAR() on the zones_online table;
fixed function ISCharacterOnlineAnywhere use now the right TIMESTAMP value from online in the character table;
if the character renamed or deleted he will automatic removed from the tables;
replace the query for SaveMainZoneOnlineTime() prevent us to showing the Zone offline if some other zone queries between this old two querys;
Special thanks to Paul to overview my Code, the suggestions and Tips;

Assambra commented:
Looks like all done now the only thing is insert on duplicate in SaveCharacterScenePath i try it, but get some error, near @'charachter' on one of the Zone Server as i remember, maybe wrong syntax; So i reset it to the old working one;

paulpach commented on Addons/NetworkZones/NetworkZones_MySql.cs:
```diff
@@ -0,0 +1,108 @@
+
+    public static bool IsCharacterOnlineAnywhere(string characterName)
+    {
+        object obj = ExecuteScalarMySql("SELECT online FROM characters WHERE name=@name", new SqlParameter("@name", characterName));
```
Try:

var obj = (MySqlDateTime)ExecuteScalarMySql(...);
if (!obj.IsNull) {
...
}

paulpach reviewed (changes requested):
don't convert date to string to date

paulpach committed:
Update and rename Readme.txt to Readme.md

Use github markdown to make this easier to read

paulpach committed:
Use primitive DateTime type

Avoid converting date time to string to date.   Use primitive date time instead

paulpach committed:
Typo

paulpach committed:
Fixed casting problem

paulpach committed:
Use DateTime

No need to transform DateTime to string to DateTime

paulpach committed:
Update NetworkZones_MySql.cs

paulpach committed:
Update NetworkZones_MySql.cs

Use insert on duplicate instead of replace into

Assambra committed:
add new version to the readme

Assambra commented:
i have tested and it works! I also increase the version for the NetworkZones addon

Assambra committed:
Merge remote-tracking branch 'upstream/master' -- sync fork with main repository --
